File: server/src/main/java/com/fluffy/auth/application/AuthService.java
package com.fluffy.auth.application;

import com.fluffy.auth.application.response.MyInfoResponse;
import com.fluffy.auth.application.response.TokenResponse;
import com.fluffy.auth.domain.Member;
import com.fluffy.auth.domain.MemberRepository;
import com.fluffy.auth.domain.OAuth2Provider;
import com.fluffy.oauth2.domain.OAuth2UserInfo;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class AuthService {

    private final TokenProvider tokenProvider;
    private final MemberRepository memberRepository;
    private final MemberMapper memberMapper;

    public Long getMemberIdByToken(String token) {
        return tokenProvider.getMemberId(token);
    }

    @Transactional(readOnly = true)
    public MyInfoResponse getMyInfo(Long memberId) {
        Member member = memberRepository.findByIdOrThrow(memberId);

        return memberMapper.toMyInfoResponse(member);
    }

    @Transactional
    public TokenResponse createToken(OAuth2UserInfo userInfo, OAuth2Provider provider) {
        Member member = findOrCreateMember(userInfo, provider);
        String accessToken = tokenProvider.createToken(member.getId().toString());

        return new TokenResponse(accessToken);
    }

    private Member findOrCreateMember(OAuth2UserInfo userInfo, OAuth2Provider provider) {
        return memberRepository.findBySocialIdAndProvider(userInfo.socialId(), provider)
                .orElseGet(() -> memberRepository.save(userInfo.toMember(provider)));
    }
}

File: server/src/main/kotlin/com/fluffy/auth/application/AuthService.kt
package com.fluffy.auth.application

import com.fluffy.auth.application.response.MyInfoResponse
import com.fluffy.auth.application.response.TokenResponse
import com.fluffy.auth.domain.Member
import com.fluffy.auth.domain.MemberRepository
import com.fluffy.auth.domain.OAuth2Provider
import com.fluffy.auth.domain.findByIdOrThrow
import com.fluffy.oauth2.domain.OAuth2UserInfo
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class AuthService(
    private val tokenProvider: TokenProvider,
    private val memberRepository: MemberRepository,
) {

    fun getMemberIdByToken(token: String): Long {
        return tokenProvider.getMemberId(token)
    }

    @Transactional(readOnly = true)
    fun getMyInfo(memberId: Long): MyInfoResponse {
        val member = memberRepository.findByIdOrThrow(memberId)

        return MemberAssembler.toMyInfoResponse(member)
    }

    @Transactional
    fun createToken(userInfo: OAuth2UserInfo, provider: OAuth2Provider): TokenResponse {
        val member = findOrCreateMember(userInfo, provider)
        val accessToken = tokenProvider.createToken(member.id.toString())

        return TokenResponse(accessToken)
    }

    private fun findOrCreateMember(userInfo: OAuth2UserInfo, provider: OAuth2Provider): Member {
        return memberRepository.findBySocialIdAndProvider(userInfo.socialId, provider)
            ?: memberRepository.save(userInfo.toMember(provider))
    }
}